Introduce WOLFSSL_DEBUG_CERTS Certificate Debug Messages

gojimmypi opened the pull request:
# Description

Adds a new troubleshooting capability to view only interesting certificate-related messages:

```c
WOLFSSL_API void WOLFSSL_MSG_CERT(const char* fmt, ...)
```

Also adds `WOLFSSL_DEBUG_LINE_ENDING` to suppress LF characters on message printed for systems that supply their own line feeds during messaging, such as the Espressif `ESP_LOG`. 

Turning on `WOLFSSL_DEBUG` _always_ enables `WOLFSSL_DEBUG_CERTS`.

However `WOLFSSL_DEBUG_CERTS` can be used _without_ `WOLFSSL_DEBUG`.

## Why?

Turning on full debugging is often overly verbose. On embedded devices the delay in printing debug messages can have an adverse effect on timing-critical code, such as certificate validation during TLS connections.

## Usage
To use, add to `user_settings.h`:

```
#define WOLFSSL_DEBUG_CERTS
```

## Sample output:

```
I (15213) time lib: Successfully set time via NTP servers.
I (15222) client-tls: get target IP address
I (15262) client-tls: Host name: www.google.com
E (15262) client-tls: Create ctx success, ret = 1
I (15262) client-tls: CA cert Use SNI success
I (15263) client-tls: Loading 8408 bytes of PEM certs
I (15272) wolfssl:      Parsed new CA
I (15273) wolfssl:      -  issuer:  '/C=US/O=Google Trust Services LLC/CN=GTS Root R1'
I (15279) wolfssl:      -  subject: '/C=US/O=Google Trust Services/CN=WR2'
I (15292) wolfssl:      Parsed new CA
I (15292) wolfssl:      -  issuer:  '/C=BE/O=GlobalSign nv-sa/OU=Root CA/CN=GlobalSign Root CA'
I (15293) wolfssl:      -  subject: '/C=US/O=Google Trust Services LLC/CN=GTS Root R1'
I (15303) wolfssl:      Parsed new CA
I (15303) wolfssl:      -  issuer:  '/C=US/O=Google Trust Services LLC/CN=GTS Root R4'
I (15303) wolfssl:      -  subject: '/C=US/O=Google Trust Services/CN=WE2'
I (15312) wolfssl:      Parsed new CA
I (15312) wolfssl:      -  issuer:  '/C=BE/O=GlobalSign nv-sa/OU=Root CA/CN=GlobalSign Root CA'
I (15312) wolfssl:      -  subject: '/C=US/O=Google Trust Services LLC/CN=GTS Root R4'
I (15325) wolfssl:      Parsed new CA
I (15325) wolfssl:      -  Failed during parse of new CA
I (15325) wolfssl:      -  error ret: -152; ASN signature error, mismatched oid
I (15329) client-tls: CA cert loaded successfully
I (15329) client-tls: Using host name: www.google.com
I (15329) client-tls: Connecting to server....www.google.com (port:443)

I (15365) client-tls: WOLFSSL object created successfully
I (15365) client-tls: wolfSSL_set_fd success
I (15365) client-tls: Connect to wolfSSL server...
```

Fixes zd# n/a

# Testing

How did you test?

Tested only manually on embedded ESP32 / ESP-IDF.

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

dgarske commented:
Jenkins retest this please: 

```
Found unhandled hudson.remoting.RequestAbortedException exception:
java.io.StreamCorruptedException: invalid stream header: 636F7272
```

gojimmypi commented on wolfcrypt/src/logging.c:
```diff
@@ -323,9 +332,25 @@ void WOLFSSL_MSG_EX2(const char *file, int line, const char* fmt, ...)
+    written = XVSNPRINTF(msg, sizeof(msg), fmt, args);
+    va_end(args);
+    if (written > 0) {
+        wolfssl_log(INFO_LOG, NULL, 0, msg);
```
Added `CERT_LOG` enum value when calling `wolfssl_log` from `WOLFSSL_MSG_CERT`.

gojimmypi commented on src/ssl.c:
```diff
@@ -5891,8 +5893,32 @@ int AddCA(WOLFSSL_CERT_MANAGER* cm, DerBuffer** pDer, int type, int verify)
+        WOLFSSL_MSG_CERT(MSG_CERT_INDENT"Failed; cert is NULL");
+    }
+#else
+    if (0) {
```
Good point. `#WOLFSSL_DEBUG_CERTS` is always on with `WOLFSSL_DEBUG` so I wrapped with `WOLFSSL_SMALL_STACK` instead.

gojimmypi commented:
After addressing items in the most recent code review from @dgarske, I added additional `WOLFSSL_MSG_CERT` messages to additional files: `asn.c`, `rsa.c`, `tfm.c`, `internal.c`

dgarske commented on src/internal.c:
```diff
@@ -15719,8 +15719,22 @@ int ProcessPeerCerts(WOLFSSL* ssl, byte* input, word32* inOutIdx,
+                        } /* ASN_NO_SIGNER_E || ASN_SELF_SIGNED_E */
+                    } /* ret != 0 && isCA */
+                #else
+                	/* When WOLFSSL_DEBUG_CERTS enabled, suggest solution */
```
```weird control chars, hard tabs, CRs, trailing whitespace:
./src/internal.c:15725:                »/* When WOLFSSL_DEBUG_CERTS enabled, suggest solution */```

gojimmypi replied:
Ugh. _Usually_ VS converts tabs to spaces automatically. Fixed.

dgarske reviewed (changes requested):
As discussed here are some patches to cleanup things.
[patch.txt](https://github.com/user-attachments/files/20840694/patch.txt)

dgarske commented on wolfcrypt/src/tfm.c:
```diff
@@ -3295,9 +3295,21 @@ int fp_exptmod_nct(fp_int * G, fp_int * X, fp_int * P, fp_int * Y)
 #if defined(WOLFSSL_ESP32_CRYPT_RSA_PRI_EXPTMOD)
    int retHW = FP_OKAY;
 #endif
+    if (P == NULL) {
```
This file has a historically different 3 space Indentation. Please have it match existing and do not refactor anything. 

Was this NULL check really needed? This should never be NULL based on calling functions. Although technically this function isn't even set as WOLFSSL_LOCAL. If you need this added its fine. 

Also I recommend adding include for logging.h at top below wolfmath.h.

gojimmypi replied:
Agree on convention; I missed the 3-space, thanks for catching that. 

I removed the NULL check, as I did not encounter a need for it. I was only concerned about this expansion:

```
#define fp_iszero(a) (((a)->used == 0) ? FP_YES : FP_NO)
```

Not an issue considering how little this file changes, re: calling functions.

Also added:  `#include <wolfssl/wolfcrypt/logging.h>`

dgarske reviewed (changes requested):
Looks great otherwise!

dgarske reviewed (dismissed):
@douzzer please finalize review

gojimmypi commented:
@dgarske I added one more [correction](https://github.com/wolfSSL/wolfssl/compare/9827b50f218b5f4a01d2e00e96b9bc7c601690b6..25878d3c78dc1f4d2c991126e63afc81b432336f#diff-34d62efe7620eac9235c878aafc6716d1ef20a2246a35fac977f34bacce0481c) and sync to upstream.

Previously I incorrectly used `minRsaKeySz` instead of `minEccKeySz` [here](https://github.com/gojimmypi/wolfssl/blob/25878d3c78dc1f4d2c991126e63afc81b432336f/src/ssl.c#L5953):

```
                    WOLFSSL_MSG_CERT("\tCA ECC key size error: pubKeySize = %d;"
                                                            " minEccKeySz = %d",
                                   cert->pubKeySize, cm->minEccKeySz);
```

Which, with `#define NO_RSA` of course caused a compile time error.

gojimmypi commented:
Jenkins retest this please

For Cannot contact wolf-linux-cloud-node-[n]: hudson.remoting.ChannelClosedException: ... Remote call on wolf-linux-cloud-node-[n] failed. The channel is closing down or has closed down

dgarske commented:
@gojimmypi please resolve new merge conflicts.

gojimmypi commented:
@dgarske @douzzer I force-pushed the branch to sync with upstream after a merge conflict developed while awaiting merge.

gojimmypi commented:
Jenkins retest this please.

For Cannot contact wolf-linux-cloud-node-[n]: java.lang.InterruptedException

douzzer reviewed (changes requested):
`g++` doesn't like some constructs here:
```
[quantum-safe-wolfssl-all-g++-latest-debug] [8 of 43] [6a5a047ee5]
    configure...   real 0m17.304s  user 0m9.606s  sys 0m9.104s
    build...609de5b546 (<jim@wolfssl.com> 2025-06-27 14:22:10 -0700 271)                         "[%s]: [%s L %d] %s"WOLFSSL_DEBUG_LINE_ENDING,
wolfcrypt/src/logging.c:271:25: error: invalid suffix on literal; C++11 requires a space between literal and string macro [-Werror=literal-suffix]
  271 |                         "[%s]: [%s L %d] %s"WOLFSSL_DEBUG_LINE_ENDING,
      |                         ^
609de5b546 (<jim@wolfssl.com> 2025-06-27 14:22:10 -0700 276)                    "[%s]: %s"WOLFSSL_DEBUG_LINE_ENDING, log_prefix, logMessage);
wolfcrypt/src/logging.c:276:20: error: invalid suffix on literal; C++11 requires a space between literal and string macro [-Werror=literal-suffix]
  276 |                    "[%s]: %s"WOLFSSL_DEBUG_LINE_ENDING, log_prefix, logMessage);
      |                    ^
609de5b546 (<jim@wolfssl.com> 2025-06-27 14:22:10 -0700 282)                         "[%s L %d] %s"WOLFSSL_DEBUG_LINE_ENDING,
wolfcrypt/src/logging.c:282:25: error: invalid suffix on literal; C++11 requires a space between literal and string macro [-Werror=literal-suffix]
  282 |                         "[%s L %d] %s"WOLFSSL_DEBUG_LINE_ENDING,
      |                         ^
609de5b546 (<jim@wolfssl.com> 2025-06-27 14:22:10 -0700 287)                         "%s"WOLFSSL_DEBUG_LINE_ENDING, logMessage);
wolfcrypt/src/logging.c:287:25: error: invalid suffix on literal; C++11 requires a space between literal and string macro [-Werror=literal-suffix]
  287 |                         "%s"WOLFSSL_DEBUG_LINE_ENDING, logMessage);
      |                         ^
cc1plus: all warnings being treated as errors
make[2]: *** [Makefile:8291: wolfcrypt/src/src_libwolfssl_la-logging.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [Makefile:10229: all-recursive] Error 1
make: *** [Makefile:5729: all] Error 2
   real 0m3.688s  user 0m23.318s  sys 0m2.071s
    scenario started 2025-06-28T02:37:58.077532Z, real elapsed 0m21.001788s
    quantum-safe-wolfssl-all-g++-latest-debug fail_build
    failed config: 'EXTRA_CPPFLAGS=-Werror' '--srcdir' '.' '--disable-jobserver' '--enable-option-checking=fatal' '--enable-all' '--enable-testcert' '--enable-acert' '--enable-dtls13' '--enable-dtls-mtu' '--enable-dtls-frag-ch' '--enable-dtlscid' '--enable-quic' '--with-sys-crypto-policy' '--enable-debug' '--enable-debug-trace-errcodes' '--enable-sp-math-all' '--enable-experimental' '--enable-kyber=yes,original' '--enable-lms' '--enable-xmss' '--enable-dilithium' '--enable-dual-alg-certs' '--disable-qt' 'CC=g++-16' 'CFLAGS=-DTEST_ALWAYS_RUN_TO_END' 'CPPFLAGS=-DNO_WOLFSSL_CIPHER_SUITE_TEST -DWOLFSSL_OLD_PRIME_CHECK'
```

gojimmypi commented:
@douzzer good catch on the g++ compiler. I'm made some adjustments, please confirm.

@dgarske no fundamental changes, just some added whitespace & explicit `WOLFSSL_DEBUG_CERTS` gating.

The reason for the new gating:  I was unable to find a robust solution for `WOLFSSL_MSG_CERT` that worked for all compilers with `WOLF_NO_VARIADIC_MACROS` enabled and `WOLFSSL_DEBUG_CERTS` disabled:

```
    #undef WOLFSSL_DEBUG_CERTS
    #ifdef WOLF_NO_VARIADIC_MACROS
        /* More than 1 param must be wrapped with WOLFSSL_DEBUG_CERTS */
        #define WOLFSSL_MSG_CERT(x)   WC_DO_NOTHING
    #else
        #define WOLFSSL_MSG_CERT(...) WC_DO_NOTHING
    #endif
```

dgarske commented on wolfssl/wolfcrypt/logging.h:
```diff
@@ -152,6 +153,28 @@ WOLFSSL_API void wolfSSL_SetLoggingPrefix(const char* prefix);
+    #undef WOLFSSL_DEBUG_CERTS
+    #ifdef WOLF_NO_VARIADIC_MACROS
+        /* More than 1 param must be wrapped with WOLFSSL_DEBUG_CERTS */
+        #define WOLFSSL_MSG_CERT(x)   WC_DO_NOTHING
```
This should be `#define WOLFSSL_MSG_CERT()   WC_DO_NOTHING`. Or are you saying that ESP32 compiler threw an error with empty ()?

dgarske replied:
Jim, you won't use `WOLF_NO_VARIADIC_MACROS` with the ESP32 compiler. Please use the suggestion I made... `#define WOLFSSL_MSG_CERT()   WC_DO_NOTHING`. It is to appease compilers that cannot handler `...` vardantic arguments. You don't need that option to pass on your compiler.

dgarske reviewed (approved):
Looks good. Over to @douzzer to finalize

gojimmypi commented:
Cool. The recent commit was only to have visibility to changes needed. I can squash upon approval.

gojimmypi commented:
Jenkins retest this please.

For ERROR: Issue with creating launcher for agent wolf-linux-cloud-node-[n]. The agent has not been fully initialized yet

gojimmypi commented on wolfcrypt/src/tfm.c:
```diff
@@ -3297,7 +3298,18 @@ int fp_exptmod_nct(fp_int * G, fp_int * X, fp_int * P, fp_int * Y)
+   }
+   if (P->used > (FP_SIZE/2)) {
+      /* FP_MAX_BITS too small is a common cert failure cause */
+#ifdef WOLFSSL_DEBUG_CERTS
```
Hi @dgarske yes, I can see the value in that. I think I've implemented what you've suggested. 

Perhaps you and @douzzer can take a look. I realize that it is an 11th hour update for the next release.

I'll continue to review & test.

gojimmypi replied:
@dgarske I've installed a WO [Open Watcom  V2 Fork](https://github.com/open-watcom/open-watcom-v2) compiler  for testing. I've implemented more robust no-variadic macro support. I'm now using no-op functions instead of variadic macros as needed.

gojimmypi commented:
Jenkins retest this please 

For AgentOfflineException: Unable to create live FilePath for wolf-linux-cloud-node-[]; wolf-linux-cloud-node-[] was marked offline: Connection was broken

dgarske reviewed (changes requested):
```
./configure CFLAGS="-DWOLF_NO_VARIADIC_MACROS -DWOLFSSL_DEBUG_CERTS" && make

wolfcrypt/src/logging.c: In function 'WOLFSSL_MSG_CERT':
wolfcrypt/src/logging.c:356:23: error: macro 'WOLFSSL_MSG_EX' requires 2 arguments, but only 1 given
  356 |     WOLFSSL_MSG_EX(msg);
      |                       ^
In file included from ./wolfssl/wolfcrypt/libwolfssl_sources.h:48,
                 from wolfcrypt/src/logging.c:22:
./wolfssl/wolfcrypt/logging.h:261:17: note: macro 'WOLFSSL_MSG_EX' defined here
  261 |         #define WOLFSSL_MSG_EX(a, b)  WC_DO_NOTHING
      |                 ^~~~~~~~~~~~~~
wolfcrypt/src/logging.c:356:5: error: 'WOLFSSL_MSG_EX' undeclared (first use in this function); did you mean 'WOLFSSL_MSG_CERT'?
  356 |     WOLFSSL_MSG_EX(msg);
      |     ^~~~~~~~~~~~~~
      |     WOLFSSL_MSG_CERT
wolfcrypt/src/logging.c:356:5: note: each undeclared identifier is reported only once for each function it appears in
wolfcrypt/src/logging.c:354:35: error: unused parameter 'msg' [-Werror=unused-parameter]
  354 | void WOLFSSL_MSG_CERT(const char* msg)
      |                       ~~~~~~~~~~~~^~~
wolfcrypt/src/logging.c: At top level:
wolfcrypt/src/logging.c:250:13: error: 'wolfssl_log' defined but not used [-Werror=unused-function]
  250 | static void wolfssl_log(const int logLevel, const char* const file_name,
      |             ^~~~~~~~~~~
wolfcrypt/src/logging.c:137:12: error: 'loggingEnabled' defined but not used [-Werror=unused-variable]
  137 | static int loggingEnabled = WOLFSSL_LOGGINGENABLED_DEFAULT;
      |            ^~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```

gojimmypi commented:
Hi @dgarske good catch. fwiw, this also fails on current master branch: 

`./configure CFLAGS="-DWOLF_NO_VARIADIC_MACROS" && make`

I'll work on getting it all fixed up.

```
./configure flags: CFLAGS=-DWOLF_NO_VARIADIC_MACROS
---
Note: Make sure your application includes "wolfssl/options.h" before any other wolfSSL headers.
      You can define "WOLFSSL_USE_OPTIONS_H" in your application to include this automatically.
make -j9  all-recursive
make[1]: Entering directory '/mnt/c/workspace/wolfssl-gojimmypi-master'
make[2]: Entering directory '/mnt/c/workspace/wolfssl-gojimmypi-master'
make[2]: warning: -j9 forced in submake: resetting jobserver mode.
  CC       wolfcrypt/test/testsuite_testsuite_test-test.o
  CC       examples/client/testsuite_testsuite_test-client.o
  CC       examples/echoclient/testsuite_testsuite_test-echoclient.o
  CC       examples/echoserver/testsuite_testsuite_test-echoserver.o
  CC       examples/server/testsuite_testsuite_test-server.o
  CC       examples/client/tests_unit_test-client.o
  CC       examples/server/tests_unit_test-server.o
  CC       wolfcrypt/test/tests_unit_test-test.o
  CC       wolfcrypt/benchmark/benchmark.o
  CC       wolfcrypt/src/src_libwolfssl_la-hmac.lo
  CC       wolfcrypt/src/src_libwolfssl_la-hash.lo
wolfcrypt/test/test.c: In function ‘ecc_test_curve’:
wolfcrypt/test/test.c:31757:58: error: macro "WOLFSSL_MSG_EX" passed 2 arguments, but takes just 0
31757 |     WOLFSSL_MSG_EX("ecc_test_curve keySize = %d", keySize);
      |                                                          ^
In file included from ./wolfssl/wolfcrypt/mem_track.h:63,
                 from wolfcrypt/test/test.c:51:
./wolfssl/wolfcrypt/logging.h:219: note: macro "WOLFSSL_MSG_EX" defined here
  219 |         #define WOLFSSL_MSG_EX()    WC_DO_NOTHING
      |
wolfcrypt/test/test.c:31757:5: error: ‘WOLFSSL_MSG_EX’ undeclared (first use in this function)
31757 |     WOLFSSL_MSG_EX("ecc_test_curve keySize = %d", keySize);
      |     ^~~~~~~~~~~~~~
wolfcrypt/test/test.c:31757:5: note: each undeclared identifier is reported only once for each function it appears in
wolfcrypt/test/test.c: In function ‘ecc_test_curve’:
wolfcrypt/test/test.c:31757:58: error: macro "WOLFSSL_MSG_EX" passed 2 arguments, but takes just 0
31757 |     WOLFSSL_MSG_EX("ecc_test_curve keySize = %d", keySize);
      |                                                          ^
In file included from ./wolfssl/wolfcrypt/mem_track.h:63,
                 from wolfcrypt/test/test.c:51:
./wolfssl/wolfcrypt/logging.h:219: note: macro "WOLFSSL_MSG_EX" defined here
  219 |         #define WOLFSSL_MSG_EX()    WC_DO_NOTHING
      |
wolfcrypt/test/test.c:31757:5: error: ‘WOLFSSL_MSG_EX’ undeclared (first use in this function)
31757 |     WOLFSSL_MSG_EX("ecc_test_curve keySize = %d", keySize);
      |     ^~~~~~~~~~~~~~
wolfcrypt/test/test.c:31757:5: note: each undeclared identifier is reported only once for each function it appears in
wolfcrypt/test/test.c: In function ‘ecc_test’:
wolfcrypt/test/test.c:34214:68: error: macro "WOLFSSL_MSG_EX" passed 2 arguments, but takes just 0
34214 |     WOLFSSL_MSG_EX("ecc_test ECC_MIN_KEY_SZ = %d\n", ECC_MIN_KEY_SZ);
      |                                                                    ^
In file included from ./wolfssl/wolfcrypt/mem_track.h:63,
                 from wolfcrypt/test/test.c:51:
./wolfssl/wolfcrypt/logging.h:219: note: macro "WOLFSSL_MSG_EX" defined here
  219 |         #define WOLFSSL_MSG_EX()    WC_DO_NOTHING
      |
wolfcrypt/test/test.c:34214:5: error: ‘WOLFSSL_MSG_EX’ undeclared (first use in this function)
34214 |     WOLFSSL_MSG_EX("ecc_test ECC_MIN_KEY_SZ = %d\n", ECC_MIN_KEY_SZ);
      |     ^~~~~~~~~~~~~~
  CC       wolfcrypt/src/src_libwolfssl_la-cpuid.lo
wolfcrypt/test/test.c: In function ‘ecc_test’:
wolfcrypt/test/test.c:34214:68: error: macro "WOLFSSL_MSG_EX" passed 2 arguments, but takes just 0
34214 |     WOLFSSL_MSG_EX("ecc_test ECC_MIN_KEY_SZ = %d\n", ECC_MIN_KEY_SZ);
      |                                                                    ^
In file included from ./wolfssl/wolfcrypt/mem_track.h:63,
                 from wolfcrypt/test/test.c:51:
./wolfssl/wolfcrypt/logging.h:219: note: macro "WOLFSSL_MSG_EX" defined here
  219 |         #define WOLFSSL_MSG_EX()    WC_DO_NOTHING
      |
wolfcrypt/test/test.c:34214:5: error: ‘WOLFSSL_MSG_EX’ undeclared (first use in this function)
34214 |     WOLFSSL_MSG_EX("ecc_test ECC_MIN_KEY_SZ = %d\n", ECC_MIN_KEY_SZ);
      |     ^~~~~~~~~~~~~~
  CC       wolfcrypt/src/src_libwolfssl_la-kdf.lo
make[2]: *** [Makefile:9993: wolfcrypt/test/testsuite_testsuite_test-test.o] Error 1
  CC       wolfcrypt/src/src_libwolfssl_la-random.lo
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [Makefile:9349: wolfcrypt/test/tests_unit_test-test.o] Error 1
make[2]: Leaving directory '/mnt/c/workspace/wolfssl-gojimmypi-master'
make[1]: *** [Makefile:10237: all-recursive] Error 1
make[1]: Leaving directory '/mnt/c/workspace/wolfssl-gojimmypi-master'
make: *** [Makefile:5718: all] Error 2
gojimmypi@DESKTOP:/mnt/c/workspace/wolfssl-gojimmypi-master$
```

dgarske commented on wolfcrypt/src/logging.c:
```diff
@@ -285,12 +375,70 @@ static void wolfssl_log(const int logLevel, const char* const file_name,
+    #include <stdarg.h> /* for var args */
+
+    #ifndef WOLFSSL_MSG_CERT_BUF_SZ
+        #define WOLFSSL_MSG_CERT_BUF_SZ 1000
```
Why such a large default? This does not conform to our small stack policy. Max 100 bytes on stack.

gojimmypi commented:
Jenkins retest this please 

For Unable to create live FilePath for wolf-linux-cloud-node-[]; wolf-linux-cloud-node-[] was marked offline: Connection was broken

JacobBarthelmeh commented:
Removing the "For this release" label. Let's hold off on merging this and have it for a feature in the next release. Thanks @gojimmypi

dgarske commented on src/ssl.c:
```diff
@@ -5893,8 +5898,35 @@ int AddCA(WOLFSSL_CERT_MANAGER* cm, DerBuffer** pDer, int type, int verify)
+    else
+    #endif
+    {
+        if (ret == 0) {
```
Just define `const char* err_msg;` here and place the `(void)err_msg;` below that at line 5926 for all cases (don't wrap in `#if defined(WOLF_NO_VARIADIC_MACROS) || defined(NO_WOLFSSL_DEBUG_CERTS)`

dgarske commented on wolfcrypt/src/logging.c:
```diff
@@ -115,22 +115,61 @@ THREAD_LS_T void *StackSizeCheck_stackOffsetPointer = 0;
 #define WOLFSSL_LOGGINGENABLED_DEFAULT 0
 #endif
 static int loggingEnabled = WOLFSSL_LOGGINGENABLED_DEFAULT;
+#ifndef WOLFSSL_CERT_LOG_ENABLED_DEFAULT
```
Can you please explain why we need/want a whole new `loggingCertEnabled`? Can we please remove this?

gojimmypi replied:
Hi @dgarske - the new `loggingCertEnabled` is a key feature of this PR: It mirrors the functionality of  the regular debugging `loggingEnabled` and the respective `wolfSSL_Debugging_ON` and `wolfSSL_Debugging_OFF`, but for certificate debug messages.  

This allows exclusive runtime enabling or disabling of the new certificate debugging messages with the new `wolfSSL_CertDebugging_ON` and `wolfSSL_CertDebugging_OFF` .

Although the new certificate debugging in this PR is a subset of the regular `DEBUG_WOLFSSL`, it can also be used as an exclusively separate debugging feature by enabling _only_ `WOLFSSL_DEBUG_CERTS` without the verbosity of full debugging.

Without the ability to turn on & off the new certificate debugging at runtime, it will otherwise only be changed with a  configuration edit and fresh compile. 

As timing of TLS connections can be critical, embedded devices displaying runtime debugging messages, sometimes as slow as 115200 or lower, the messages can introduce new problems when debugging. The granular adjustment of message volume would be a valuable feature.

Given this information, are you willing to consider keeping it? Please let me know more about your aversion to this.

gojimmypi commented:
The latest commit (squashed, forced push) introduces a new changes to the wolfSSL `CMakeLists.txt`:
- During `WOLFSSL_DEBUG` the `-g` is not supported in Watcom. This is now properly detected and handled.

- While there, a new, optional section was added to inspect all cmake variables during debug. 

The `logging.h` was revised to not define `__func__` for Watcom compilers during `WOLFSSL_DEBUG` . Otherwise:

```
Warning! W140: Definition of macro  '__func__' not identical to previous definition
```

Changes made to `src/ssl.c` per [code review](https://github.com/wolfSSL/wolfssl/pull/8902#discussion_r2220025538). See my [explanation](https://github.com/wolfSSL/wolfssl/pull/8902#discussion_r2220210442) to retain proposed `loggingCertEnabled` functionality.

dgarske reviewed (changes requested):
This is does add a nice way to log just certificates from the peer. 
I think the code size increase is small and you did provide a way via NO_WOLFSSL_DEBUG_CERTS to disable it.

gojimmypi commented on wolfcrypt/src/logging.c:
```diff
@@ -115,22 +115,61 @@ THREAD_LS_T void *StackSizeCheck_stackOffsetPointer = 0;
+        loggingCertEnabled = 1;
+    #endif
+    #if defined(WOLFSSL_APACHE_MYNEWT)
+        log_register("wolfcrypt", &mynewt_log, &log_console_handler,
```
> Please confirm calling `log_register` twice is okay...

That's a good observation! It's probably _not_ a good idea to call it more than once. 

In fact, even without this PR, it was likely always problematic with multiple `wolfSSL_Debugging_ON` calls.

See docs: https://mynewt.apache.org/latest/os/modules/logs/logs.html 

Although there's nothing explicit in the docs, and I too cannot easily confirm, I had this response from an AI:

---

### Safety of Calling `log_register()` Multiple Times

Calling `log_register()` more than once on the same `struct log` instance **is not safe** and can lead to issues:

- **Duplicate registration** will cause `log_register()` to return an error, typically `SYS_EALREADY` or `SYS_EINVAL`.
- **Re-registering with a different handler or log level has no effect** — the existing registration remains unchanged.
- **No automatic override**: It does **not** update or replace the log configuration if already registered.
- **Internals**: The Mynewt logging system maintains a global registry of log instances and expects each to be registered once.

---

### Action Taken

Although it sounds believable, I did not confirm.  Regardless of how true the AI response may be or not, it is a relatively simple process (and probably a Best Practice) to ensure `log_register` is called only once, not only for repeated calls to `wolfSSL_Debugging_ON`, but also the new `wolfSSL_CertDebugging_ON`.

### Simulated Apache Mynewt

The logic was tested, but not with the actual Apache Mynewt,

To test this, I commented out [these lines](https://github.com/wolfSSL/wolfssl/blob/8f00fc25946d722fd472a475c3b5c3ef07c6f6cb/wolfcrypt/src/logging.c#L128) 

```c
#if defined(WOLFSSL_APACHE_MYNEWT)
//  #include "log/log.h"
// static struct log mynewt_log;
#endif /* WOLFSSL_APACHE_MYNEWT */
```

And added this  _after_ the `libwolfssl_sources.h` at the top of the `logging.c` file:

```c
/* TODO REMOVE SECTION */
#define WOLFSSL_APACHE_MYNEWT

#if defined(WOLFSSL_APACHE_MYNEWT)
    static int log_console_handler = 0;
    static int LOG_SYSLEVEL = 0;
    static int log_register(char* src,  void* a, void* b, void* c, int d)
    {
        return 0;
    }
    typedef struct
    {
        int a;
    } log;
    static  log mynewt_log;
#endif
/* END TODO REMOVE */
```

Then test with this (manual inspection):

```c
    wolfSSL_Debugging_ON();
    wolfSSL_Debugging_ON();
    wolfSSL_CertDebugging_ON();
    wolfSSL_CertDebugging_ON();
    wolfSSL_CertDebugging_OFF();
    wolfSSL_Debugging_ON();
    wolfSSL_Debugging_OFF();
    wolfSSL_Debugging_ON();
    wolfSSL_CertDebugging_ON();
```

gojimmypi commented:
Jenkins retest this please

for wolf-linux-cloud-node-[n] is offline

douzzer reviewed (changes requested):
have you looked at what `WOLFSSL_DEBUG_CERTIFICATE_LOADS` does?
```
$ git grep WOLFSSL_DEBUG_CERTIFICATE_LOADS
README.md:* Added WOLFSSL_DEBUG_PRINTF and WOLFSSL_DEBUG_CERTIFICATE_LOADS for improved debugging output. (PR #8769, PR #8770)
src/ssl_load.c:#ifdef WOLFSSL_DEBUG_CERTIFICATE_LOADS
src/ssl_load.c:#ifdef WOLFSSL_DEBUG_CERTIFICATE_LOADS
src/ssl_load.c:#endif /* WOLFSSL_DEBUG_CERTIFICATE_LOADS */
src/ssl_load.c:#ifdef WOLFSSL_DEBUG_CERTIFICATE_LOADS
src/ssl_load.c:#endif /* WOLFSSL_DEBUG_CERTIFICATE_LOADS */
src/ssl_load.c:#ifdef WOLFSSL_DEBUG_CERTIFICATE_LOADS
```

it looks like you're getting us a lot more coverage with this PR, but I don't think it makes sense to have both `WOLFSSL_DEBUG_CERTIFICATE_LOADS` and `WOLFSSL_DEBUG_CERTS`.  at a minimum we should settle on a consistent naming scheme for the feature macros that either merges them into one, or gives them names that make clear what's different between them.

gojimmypi commented:
Hi @douzzer - yes, I recently noticed the (relatively new?) `WOLFSSL_DEBUG_CERTIFICATE_LOADS` and  `WOLFSSL_DEBUG_PRINTF`.

I've modelled the `WOLFSSL_DEBUG_CERTS`  in this PR after the existing `WOLFSSL_DEBUG`.

I'm curious why `WOLFSSL_DEBUG_PRINTF` was introduced when there's been `WOLFSSL_MSG` and `WOLFSSL_MSG_EX` widely in use?

There are relatively few instances of `WOLFSSL_DEBUG_CERTIFICATE_LOAD`  (only 4 total, and only in `ssl_load.c`).

Is there any pressing reason to have `WOLFSSL_DEBUG_CERTIFICATE_LOAD` separate?

Any objection to have the `WOLFSSL_DEBUG_CERTIFICATE_LOAD` merged into `WOLFSSL_DEBUG_CERTS`? I can do that either in this PR, or a future PR.

I already have more certificate debugging in my dev branch, this PR was mainly to introduce the concept.

In any case, I'm glad you also had the good idea to have some addition, certificate-specific debugging. 

Please let me know you preference on how to proceed.

dgarske commented:
Please resolve merge conflicts. Please see if changing to a simplier WOLFSSL_DEBUG_CERTIFICATE_LOADS makes more sense. If not then please cleanup WOLFSSL_DEBUG_CERTIFICATE_LOADS.

gojimmypi commented:
Hi @dgarske and @douzzer - 

There were a small number of `WOLFSSL_DEBUG_CERTIFICATE_LOADS `. I propose changing them to `WOLFSSL_DEBUG_CERTS`.

I applied this change in https://github.com/wolfSSL/wolfssl/pull/8902/commits/7020b3aba99f2724f528d683ad26b9fb87b42843, and left a note in `logging.h`:

```
/* WOLFSSL_DEBUG_PRINTF_FN is intended to be used only in wolfssl_log(),
 * but is exposed in header as a customer cross-platform debugging capability.
 *
 * All general wolfSSL debugging should use:
 *   WOLFSSL_MSG and WOLFSSL_MSG_EX
 *
 * All wolfSSL certificate-related debugging should use:
 *   WOLFSSL_MSG_CERT and WOLFSSL_MSG_CERT_EX
 *
 * For custom debugging output, define your own WOLFSSL_DEBUG_PRINTF_FN
 */
```

dgarske commented:
Jenkins retest this please: "AgentOfflineException"

douzzer reviewed (dismissed):
Please do not substitute `WOLFSSL_MSG_CERT_EX()` for `WOLFSSL_DEBUG_PRINTF()`.

`WOLFSSL_DEBUG_PRINTF()` is a bufferless, non-truncating debug message renderer.

The refactor in ssl_load.c to use `WOLFSSL_MSG_CERT_EX()` causes destructive message truncation.

Also, there are merge conflicts on logging.c.

gojimmypi commented:
Jenkins retest this please.

For AgentOfflineException: Unable to create live FilePath for wolf-linux-cloud-node-[]; wolf-linux-cloud-node-[] was marked offline: Connection was broken

gojimmypi commented:
> Please do not substitute WOLFSSL_MSG_CERT_EX() for WOLFSSL_DEBUG_PRINTF().

Understood. I've reverted that change.

I don't understand why we need a new debugging message tool in addition to `WOLFSSL_MSG` and `WOLFSSL_MSG_EX`.  Those could have been modified to be optionally bufferless & unlimited lengths. `WOLFSSL_DEBUG_PRINTF()` also cannot be enabled/disabled, or queried.

> `WOLFSSL_DEBUG_PRINTF()` is a bufferless, non-truncating debug message renderer.
> The refactor in ssl_load.c to use WOLFSSL_MSG_CERT_EX() causes destructive message truncation.

I've added comments to the `logging.h` regarding `WOLFSSL_DEBUG_PRINTF` attributes.

There's of course `WOLFSSL_MSG_BUF_SZ` and the new `WOLFSSL_MSG_CERT_BUF_SZ` than can be assigned for larger messages.

> Also, there are merge conflicts on logging.c.

This must have been since resolved as I didn't see it, but in any case I've refreshed my branch from upstream.

dgarske commented on wolfcrypt/src/poly1305.c:
```diff
@@ -88,15 +88,17 @@ static word32 cpu_flags_set = 0;
 #if defined(USE_INTEL_POLY1305_SPEEDUP) || defined(POLY130564)
-    #if defined(_MSC_VER)
+    #if defined(__WATCOMC__)
+        #error "POLY130564 || USE_INTEL_POLY1305_SPEEDUP Watcom not supported"
```
Do we really need to throw an error here?

gojimmypi replied:
Yes, the alternative is a very unintuitive error.

Example config:

```
cmake -B build-watcom -G "Watcom WMake" ^
  -DCMAKE_C_FLAGS="-DWOLFSSL_DLL -DBUILDING_WOLFSSL -DPOLY130564 " ^
  -DCMAKE_SYSTEM_NAME=Windows ^
  -DCMAKE_EXECUTABLE_SUFFIX=.exe ^
  -DCMAKE_SYSTEM_PROCESSOR=i386 ^
  -DCMAKE_VERBOSE_MAKEFILE=TRUE ^
  -DWOLFSSL_ASM=NO ^
  -DWOLFSSL_EXAMPLES=YES ^
  -DWOLFSSL_CRYPT_TESTS=NO ^
  -DWOLFSSL_SINGLE_THREADED=YES ^
  -DBUILD_SHARED_LIBS=NO ^
  -DCMAKE_POLICY_DEFAULT_CMP0136=NEW ^
  -DCMAKE_WATCOM_RUNTIME_LIBRARY=SingleThreaded
if errorlevel 1 (
    echo "Config 2 failed!"
    exit /b
)
cmake --build build-watcom --target client
```

results in this error:

```
Open Watcom C/C++ x86 32-bit Compile and Link Utility
Version 2.0 beta Jul 11 2025 02:52:46 (32-bit)
Copyright (c) 2002-2025 The Open Watcom Contributors. All Rights Reserved.
Portions Copyright (c) 1988-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.
See https://github.com/open-watcom/open-watcom-v2#readme for details.
        wcc386 C:\workspace\wolfssl-pr-cert\wolfcrypt\src\poly1305.c  -d+ -DBUILDING_WOLFSSL -DECC_SHAMIR -DECC_TIMING_RESISTANT -DGCM_TABLE_4BIT -DHAVE_AESGCM -DHAVE_CHACHA -DHAVE_CONFIG_H -DHAVE_DH_DEFAULT_PARAMS -DHAVE_ECC -DHAVE_ENCRYPT_THEN_MAC -DHAVE_EXTENDED_MASTER -DHAVE_FFDHE_2048 -DHAVE_HASHDRBG -DHAVE_HKDF -DHAVE_ONE_TIME_AUTH -DHAVE_POLY1305 -DHAVE_SUPPORTED_CURVES -DHAVE_THREAD_LS -DHAVE_TLS_EXTENSIONS -DNO_CHACHA_ASM -DNO_DES3 -DNO_DES3_TLS_SUITES -DNO_DSA -DNO_MD4 -DNO_MD5 -DNO_OLD_TLS -DNO_PSK -DNO_RC4 -DSINGLE_THREADED -DTFM_ECC256 -DTFM_NO_ASM -DTFM_TIMING_RESISTANT -DWC_NO_ASYNC_THREADING -DWC_RSA_BLINDING -DWC_RSA_PSS -DWOLFSSL_HAVE_MIN -DWOLFSSL_HAVE_MAX -DNO_WRITEV -DWOLFSSL_IGNORE_FILE_WARN -DWOLFSSL_NO_ASM -DWOLFSSL_NO_SHAKE128 -DWOLFSSL_NO_SHAKE256 -DWOLFSSL_PSS_LONG_SALT -DWOLFSSL_SHA384 -DWOLFSSL_SHA512 -DWOLFSSL_SYS_CA_CERTS -DWOLFSSL_TLS13 -DWOLFSSL_USE_ALIGN -I"C:\workspace\wolfssl-pr-cert" -I"C:\workspace\wolfssl-pr-cert\build-watcom" -I"C:\watcom\h" -I"C:\watcom\h\nt" -DWOLFSSL_DLL -DBUILDING_WOLFSSL -DPOLY130564 -wx -wcd=202 -d2 -foCMakeFiles\wolfssl.dir\wolfcrypt\src\poly1305.c.obj
Open Watcom C x86 32-bit Optimizing Compiler
Version 2.0 beta Jul 11 2025 02:49:30 (32-bit)
Copyright (c) 2002-2025 The Open Watcom Contributors. All Rights Reserved.
Portions Copyright (c) 1984-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.
See https://github.com/open-watcom/open-watcom-v2#readme for details.
C:\workspace\wolfssl-pr-cert\wolfcrypt\src\poly1305.c(288): Error! E1011: Symbol 'word128' has not been declared
C:\workspace\wolfssl-pr-cert\wolfcrypt\src\poly1305.c(288): Error! E1009: Expecting ';' but found 'd0'
C:\workspace\wolfssl-pr-cert\wolfcrypt\src\poly1305.c(288): Error! E1011: Symbol 'd0' has not been declared
C:\workspace\wolfssl-pr-cert\wolfcrypt\src\poly1305.c(288): Error! E1011: Symbol 'd1' has not been declared
C:\workspace\wolfssl-pr-cert\wolfcrypt\src\poly1305.c(288): Error! E1011: Symbol 'd2' has not been declared
C:\workspace\wolfssl-pr-cert\wolfcrypt\src\poly1305.c(288): Error! E1011: Symbol 'd' has not been declared
C:\workspace\wolfssl-pr-cert\wolfcrypt\src\poly1305.c(313): Warning! W131: No prototype found for function 'MUL'
C:\workspace\wolfssl-pr-cert\wolfcrypt\src\poly1305.c(313): Warning! W131: No prototype found for function 'ADD'
C:\workspace\wolfssl-pr-cert\wolfcrypt\src\poly1305.c(318): Warning! W131: No prototype found for function 'SHR'
C:\workspace\wolfssl-pr-cert\wolfcrypt\src\poly1305.c(318): Warning! W131: No prototype found for function 'LO'
C:\workspace\wolfssl-pr-cert\wolfcrypt\src\poly1305.c(319): Warning! W131: No prototype found for function 'ADDLO'
C:\workspace\wolfssl-pr-cert\wolfcrypt\src\poly1305.c(313): Warning! W308: The function 'MUL' without prototyped parameters called
C:\workspace\wolfssl-pr-cert\wolfcrypt\src\poly1305.c(313): Note! N2002: 'MUL' defined in: C:\workspace\wolfssl-pr-cert\wolfcrypt\src\poly1305.c(313)

[  ... snip similar errors ... ]

C:\workspace\wolfssl-pr-cert\wolfcrypt\src\poly1305.c(320): Note! N2002: 'LO' defined in: C:\workspace\wolfssl-pr-cert\wolfcrypt\src\poly1305.c(318)
C:\workspace\wolfssl-pr-cert\wolfcrypt\src\poly1305.c: 1080 lines, included 153253, 28 warnings, 6 errors
Error: Unable to invoke "C:\watcom\binnt\wcc386.exe"
Error(E42): Last command making (CMakeFiles\wolfssl.dir\wolfcrypt\src\poly1305.c.obj) returned a bad status
Error(E02): Make execution terminated
Error(E42): Last command making (CMakeFiles\wolfssl.dir\all) returned a bad status
Error(E02): Make execution terminated
Error(E42): Last command making (CMakeFiles\client.dir\rule) returned a bad status
Error(E02): Make execution terminated
Error(E42): Last command making (client) returned a bad status
Error(E02): Make execution terminated
```

It might be resolvable, but I didn't chase it for this PR.

dgarske commented on src/ssl.c:
```diff
@@ -5903,7 +5931,7 @@ int AddCA(WOLFSSL_CERT_MANAGER* cm, DerBuffer** pDer, int type, int verify)
 
     /* check CA key size */
-    if (verify) {
+    if (verify && (ret == 0 )) {
```
Nice find here! Doesn't cause any harm as it was, but can cause the original error code to be overridden.

dgarske commented on wolfssl/wolfcrypt/logging.h:
```diff
@@ -152,8 +241,50 @@ WOLFSSL_API void wolfSSL_SetLoggingPrefix(const char* prefix);
+    WOLFSSL_API int WOLFSSL_MSG_CERT_EX(const char* fmt, ...);
+#else
+    /* No Certificate Debugging */
+    #ifndef WOLFSSL_MSG_CERT_INDENT
```
Please make sure `WOLFSSL_DEBUG_CERTS` is undefined.

gojimmypi commented on wolfssl/wolfcrypt/logging.h:
```diff
@@ -26,11 +26,94 @@
+ * WOLFSSL_DEBUG_PRINTF_FN(...)
+ *  A buffer-less, non-truncating debug message renderer.
+ *
+ * WOLFSSL_MSG_BUF_SZ
```
oops, good catch. 

`WOLFSSL_MSG_EX_BUF_SZ` is used only by `WOLFSSL_MSG_EX()`.

There's no size for `WOLFSSL_MSG()`, which just calls non-variadic parameter [`wolfssl_log(INFO_LOG, NULL, 0, msg)`](https://github.com/wolfSSL/wolfssl/blob/a1e2ba2cd33e20eb63e0e857f9304122e429efcf/wolfcrypt/src/logging.c#L334)

Unless otherwise advised, I have removed that comment.

gojimmypi commented on wolfcrypt/src/logging.c:
```diff
@@ -323,16 +468,24 @@ void WOLFSSL_MSG_EX2(const char *file, int line, const char* fmt, ...)
+#endif /* WOLFSSL_DEBUG_CODEPOINTS */
 
-#endif
+#else
```
Documentation. 

Although often best in a comment, I thought it would be a better illustration.

I found the entire exercise of non-variadics rather unintuitive. This section is for my future self, or someone else that might go digging here.

The key point is that there's no implementation in the `logging.c` file during the `else`, and the placeholder varies depending on whether variadic macros are supported or not (e.g. Watcom).  I'd like to leave this for the resultant syntax-highlighting that can aid in understanding:

```c
#else
    /* We need a do-nothing function when variadic macros not available */
    #ifdef WOLF_NO_VARIADIC_MACROS
        /* See header for same-name, do nothing functions */
    #else
        /* See header for same-name, do nothing macros */
    #endif
#endif /* XVSNPRINTF && !NO_WOLFSSL_MSG_EX */
```

It was however, missing the do-nothing function comment that will be included in my next update unless otherwise advised.

gojimmypi commented on wolfcrypt/src/logging.c:
```diff
@@ -285,12 +352,89 @@ static void wolfssl_log(const int logLevel, const char* const file_name,
+            wolfssl_log(INFO_LOG, NULL, 0, msg);
+        }
+#ifdef WOLFSSL_SMALL_STACK
+    XFREE(msg, NULL, DYNAMIC_TYPE_OUT_BUFFER);
```
both instances changed

gojimmypi commented on src/ssl.c:
```diff
@@ -5869,7 +5869,7 @@ int AddCA(WOLFSSL_CERT_MANAGER* cm, DerBuffer** pDer, int type, int verify)
     DerBuffer*   der = *pDer;
 
-    WOLFSSL_MSG("Adding a CA");
+    WOLFSSL_MSG_CERT("Adding a CA");
```
This is an interesting comment.

The premise question: _Does turning off *certificate* debugging turn off ALL certificate-related debug messages?_

Originally, I thought the answer would be yes. No certificate debugging. Don't display it.

But another premise is that: _Certificate debugging is a *subset* of regular debugging_.

Reminder that `WOLFSSL_MSG_CERT("Adding a CA");` is the same as:

```
#ifdef WOLFSSL_DEBUG_CERTS
WOLFSSL_MSG_CERT("Adding a CA");
#endif
```

Your logic of course overrides the premise, and always displays the "`Adding a CA`" message when debugging: cert debugging or not.

I need to give some more thought to it. Open to any additional comments. 

Your thoughts here are that some certificate-related debug messages are simply too important to turn off?

gojimmypi replied:
Here's my proposed logic for showing debug messages for both regular (MSG) and certificate (CERT).

### MSG: `WOLFSSL_MSG` and `WOLFSSL_MSG_EX`
### CERT: `WOLFSSL_MSG_CERT` and `WOLFSSL_MSG_CERT_EX`


| **Debug** | **No Cert Debug** | **Cert Debug**  | **MSG**  | **CERT** |
|-------------------- |---------------------------------|--------------------------------|------|-------|
|                              |                                                  |                                             | ----- | ----- |
|                              |                                                  | WOLFSSL_DEBUG_CERTS    | ----- | show  |
|                              | NO_WOLFSSL_DEBUG_CERTS  |                                             | ----- |  -----  |
|                              | NO_WOLFSSL_DEBUG_CERTS  | WOLFSSL_DEBUG_CERTS    | ----- | ----- |
| WOLFSSL_DEBUG |                                                 |                                             | show | show |
| WOLFSSL_DEBUG |                                                  | WOLFSSL_DEBUG_CERTS    | show | show  |
| WOLFSSL_DEBUG | NO_WOLFSSL_DEBUG_CERTS  |                                             | show | ----- |
| WOLFSSL_DEBUG | NO_WOLFSSL_DEBUG_CERTS  | WOLFSSL_DEBUG_CERTS    | show | ----- |

gojimmypi commented on src/crl.c:
```diff
@@ -791,7 +792,7 @@ int BufferLoadCRL(WOLFSSL_CRL* crl, const byte* buff, long sz, int type,
     crl->currentEntry = CRL_Entry_new(crl->heap);
     if (crl->currentEntry == NULL) {
-        WOLFSSL_MSG("alloc CRL Entry failed");
+        WOLFSSL_MSG_CERT("alloc CRL Entry failed");
```
As noted in the [thread above](https://github.com/wolfSSL/wolfssl/pull/8902#discussion_r2241303761) I've introduced a new macro `WOLFSSL_MSG_CERT_LOG` that implements your [proposed logic](https://github.com/wolfSSL/wolfssl/pull/8902#discussion_r2240759818):

```c
#ifdef WOLFSSL_DEBUG_CERTS
WOLFSSL_MSG("Adding a CA");
#else
WOLFSSL_MSG_CERT("Adding a CA");
#endif
```

dgarske commented:
Jenkins retest this please: PRB-dtls.txt_27 failed with " ClienSSL_write msg error -397, Peer closed underlying transport Error". Likely UDP issue.

dgarske commented:
@JacobBarthelmeh or @douzzer can you please review this as well and merge if you approve? Thank you

douzzer commented on src/ssl_load.c:
```diff
@@ -2721,57 +2724,65 @@ int ProcessFile(WOLFSSL_CTX* ctx, const char* fname, int format, int type,
         else if (wc_PemGetHeaderFooter(CERT_TYPE, &header, &footer) == 0 &&
                 (XSTRNSTR((char*)content.buffer, header, (word32)sz) !=
                     NULL)) {
             type = CERT_TYPE;
+            WOLFSSL_DEBUG_PRINTF("Detected cert type CERT_TYPE = %d:", type);
         }
         else
-#endif
+#endif /* !NO_CODING && !WOLFSSL_NO_PEM */
         {
             /* Not a header that we support. */
-            WOLFSSL_MSG("Failed to detect certificate type");
-#ifdef WOLFSSL_DEBUG_CERTIFICATE_LOADS
+            WOLFSSL_MSG_CERT_LOG("Failed to detect certificate type");
             WOLFSSL_DEBUG_PRINTF(
                 "ERROR: ProcessFile: Failed to detect certificate type"
                 " of \"%s\"\n",
                 fname);
```
`WOLFSSL_DEBUG_PRINTF()` is unavailable on some targets, and has no default no-op definition, so the `WOLFSSL_DEBUG_CERTIFICATE_LOADS` gate around it is needed to avoid breaking buildability on those targets.  You'll need to either restore the gate (if you gate on `WOLFSSL_DEBUG_CERTS` then `WOLFSSL_DEBUG_CERTS` will be limited to targets that have `WOLFSSL_DEBUG_PRINTF()` defined), or you'll need to add a fallthrough no-op definition of `WOLFSSL_DEBUG_PRINTF()`.

Either way, bear in mind that `WOLFSSL_DEBUG_PRINTF()` is available on supported targets even in builds with `WOLFSSL_DEBUG` undefined, to allow narrowly scoped debugging builds, as used with `WOLFSSL_DEBUG_CERTIFICATE_LOADS` and the `all-check-certs-one-year` scenario on `wolfssl-multi-test.sh` that covers us for in-repo expiring certs with logging of the path to the cert(s) that will expire soon.

gojimmypi replied:
Thanks for the clarification. I've reverted the `#ifdef WOLFSSL_DEBUG_CERTIFICATE_LOADS` for now.

> unavailable on some targets, and has no default no-op definition

It's unfortunate that those targets are not tested. 

If you provide an example of the problematic targets, I'll create a GitHub workflow config to check it.

douzzer commented on tests/api.c:
```diff
@@ -19884,6 +19884,8 @@ static int test_wc_i2d_PKCS12(void)
     return EXPECT_RESULT();
 }
 
+#ifndef __WATCOMC__
+/* There are known Watcom incompatibilities in this implementation  */
```
How did this off-topic Watcom compiler stuff sneak into this PR?  Seems like we should fix this with higher priority in a separate PR.

gojimmypi replied:
Upon reconsideration, it was inappropriate for me to include the Watcom gate-out to avoid the error. 

I've excluded that update from the most recent commit / squash in this PR.

I still plan to create a new PR to address the watcom compiler issue, including a new test since it is not detected by any current tests.

douzzer commented on wolfssl/wolfcrypt/logging.h:
```diff
@@ -26,11 +26,95 @@
+ * Message / printf debugging
+ * ****************************************************************************
+ *
+ * WOLFSSL_DEBUG_PRINTF_FN(...)
+ *  A buffer-less, non-truncating debug message renderer.
```
`WOLFSSL_DEBUG_PRINTF()` is the utility macro.  `WOLFSSL_DEBUG_PRINTF_FN` is used to supply an override definition of the target platform's `printf`-like function, and it is not function-like:
```
#ifdef WOLFSSL_DEBUG_PRINTF_FN
    /* user-supplied definition */
#elif defined(ARDUINO)
    /* ARDUINO only has print and sprintf, no printf. */
#elif defined(WOLFSSL_LOG_PRINTF) || defined(WOLFSSL_DEOS)
    #define WOLFSSL_DEBUG_PRINTF_FN printf
[...]
```

gojimmypi replied:
Ah, excellent. I'll update the documentation comments in logging.h - thanks for the clarification.

gojimmypi commented:
Pushed a new [commit](https://github.com/wolfSSL/wolfssl/compare/9cfbb7791867d2860babba5721ebed30ebfacf7f..d64ef34ef89c2734e1e2c9b7551368c80014dfc2) since the embedded `//` in a block comment was flagged as problem:

```
/*
 ...
 *   #ifdef WOLFSSL_DEBUG_PRINTF_FN
 *       #define [user-supplied definition]
 *   #elif defined(ARDUINO)
 *      // ARDUINO only has print and sprintf, no printf on some targets.
 *   #elif defined(WOLFSSL_LOG_PRINTF) || defined(WOLFSSL_DEOS)
 *       #define WOLFSSL_DEBUG_PRINTF_FN printf
 *   [...]
 */
```

douzzer commented on .github/workflows/os-check.yml:
```diff
@@ -57,7 +57,8 @@ jobs:
           '--enable-lms=small,verify-only --enable-xmss=small,verify-only',
-          '--disable-sys-ca-certs'
+          '--disable-sys-ca-certs',
+          '--enable-all CPPFLAGS=-DWOLFSSL_DEBUG_CERTS ',
```
Never mind, I misread the code snippet.  I wrongly assumed I was looking at an array constructor, but this was a matrix in a yaml file.  All good.
